File: src/components/agent/session-tree.ts
// Sidebar session ordering — pure, shared by the AgentTab list render AND the
// j/k cycling in the store, so "next/prev session" matches what's on screen.
//
// Order = sessions clustered by cwd into a group tree (ordered within + across
// groups by the persisted flat `sessionOrder`), with fork lineage nesting each
// fork right after its parent. `flattenSidebarOrder` walks that tree in render
// order to produce the flat list a user sees top-to-bottom.

import type { SessionInfo } from '@/store/agent'

export interface GroupNode {
  cwd: string                  // '' = "no directory" bucket
  label: string                // path segment relative to parent
  fullPath: string             // for tooltip
  sessions: SessionInfo[]
  children: GroupNode[]
  depth: number
}

function dirBasename(path: string): string {
  const parts = path.replace(/\/+$/, '').split('/')
  return parts[parts.length - 1] || path
}

export function buildGroupTree(sessions: SessionInfo[], order: string[]): GroupNode[] {
  const orderIdx = new Map(order.map((id, i) => [id, i]))
  const sessionKey = (s: SessionInfo) => orderIdx.get(s.id) ?? Number.MAX_SAFE_INTEGER
  const sortSessions = (a: SessionInfo, b: SessionInfo) => {
    const ai = sessionKey(a)
    const bi = sessionKey(b)
    if (ai !== bi) return ai - bi
    return b.createdAt - a.createdAt
  }

  // Bucket by cwd
  const byCwd = new Map<string, SessionInfo[]>()
  for (const s of sessions) {
    const cwd = s.cwd ?? ''
    const arr = byCwd.get(cwd) ?? []
    arr.push(s)
    byCwd.set(cwd, arr)
  }
  for (const arr of byCwd.values()) arr.sort(sortSessions)

  // Sort cwds shortest-first so each node's parent already exists when we link
  const cwds = [...byCwd.keys()].sort((a, b) => a.length - b.length)
  const nodeByCwd = new Map<string, GroupNode>()
  for (const cwd of cwds) {
    nodeByCwd.set(cwd, {
      cwd,
      label: cwd ? dirBasename(cwd) : '(no directory)',
      fullPath: cwd,
      sessions: byCwd.get(cwd)!,
      children: [],
      depth: 0,
    })
  }

  const roots: GroupNode[] = []
  for (const cwd of cwds) {
    const node = nodeByCwd.get(cwd)!
    // Find longest *sessioned* cwd that is a strict ancestor
    let parentCwd: string | null = null
    if (cwd) {
      for (const other of cwds) {
        if (!other || other === cwd) continue
        if (cwd.startsWith(other + '/') && (parentCwd === null || other.length > parentCwd.length)) {
          parentCwd = other
        }
      }
    }
    if (parentCwd !== null) {
      const parent = nodeByCwd.get(parentCwd)!
      parent.children.push(node)
      node.depth = parent.depth + 1
      node.label = cwd.slice(parentCwd.length + 1)
    } else {
      roots.push(node)
    }
  }

  // Group sort key: first-member's order index, recursing into children if empty
  const groupKey = (n: GroupNode): number => {
    if (n.sessions.length > 0) return sessionKey(n.sessions[0]!)
    if (n.children.length > 0) return groupKey(n.children[0]!)
    return Number.MAX_SAFE_INTEGER
  }
  const sortGroupsRec = (arr: GroupNode[]) => {
    arr.sort((a, b) => groupKey(a) - groupKey(b))
    for (const n of arr) sortGroupsRec(n.children)
  }
  sortGroupsRec(roots)

  return roots
}

/** If the tree has a single root (one cwd shared by everything), drop its
 *  redundant header — promote its sessions to the top level (alongside Al)
 *  and its child groups become the new roots. */
export function peelUniversalRoot(roots: GroupNode[]): { rootSessions: SessionInfo[]; roots: GroupNode[] } {
  if (roots.length !== 1) return { rootSessions: [], roots }
  const only = roots[0]!
  const promoted = only.children.map((c) => shiftDepth(c, -1))
  return { rootSessions: only.sessions, roots: promoted }
}

function shiftDepth(node: GroupNode, delta: number): GroupNode {
  return {
    ...node,
    depth: node.depth + delta,
    children: node.children.map((c) => shiftDepth(c, delta)),
  }
}

/** Arrange a flat list of sessions (all sharing one cwd group) into a fork
 *  lineage: each fork is emitted right after its parent, one indent deeper.
 *  Preserves the incoming order for roots; a fork whose parent isn't in this
 *  list is treated as a root. */
export function arrangeLineage(sessions: SessionInfo[]): Array<{ session: SessionInfo; depth: number }> {
  const inSet = new Set(sessions.map((s) => s.claudeSessionId).filter(Boolean) as string[])
  const childrenOf = new Map<string, SessionInfo[]>()
  for (const s of sessions) {
    const p = s.parentClaudeSessionId
    if (p && inSet.has(p)) {
      const arr = childrenOf.get(p) ?? []
      arr.push(s)
      childrenOf.set(p, arr)
    }
  }
  const out: Array<{ session: SessionInfo; depth: number }> = []
  const emit = (s: SessionInfo, depth: number) => {
    out.push({ session: s, depth })
    if (s.claudeSessionId) {
      for (const child of childrenOf.get(s.claudeSessionId) ?? []) emit(child, depth + 1)
    }
  }
  for (const s of sessions) {
    const isRoot = !s.parentClaudeSessionId || !inSet.has(s.parentClaudeSessionId)
    if (isRoot) emit(s, 0)
  }
  return out
}

/** The flat, top-to-bottom order of sessions as the sidebar renders them
 *  (excludes Al — callers prepend it). Collapsed groups are skipped (their
 *  sessions + children are hidden in the UI), so cycling matches what's visible. */
export function flattenSidebarOrder(sessions: SessionInfo[], order: string[], collapsedGroups?: Set<string>): SessionInfo[] {
  const { rootSessions, roots } = peelUniversalRoot(buildGroupTree(sessions, order))
  const out: SessionInfo[] = []
  for (const { session } of arrangeLineage(rootSessions)) out.push(session)
  const walk = (nodes: GroupNode[]) => {
    for (const n of nodes) {
      if (collapsedGroups?.has(n.cwd)) continue
      for (const { session } of arrangeLineage(n.sessions)) out.push(session)
      walk(n.children)
    }
  }
  walk(roots)
  return out
}

File: android/app/src/main/kotlin/io/amar/console/ui/agents/SessionTree.kt
package io.amar.console.ui.agents

import io.amar.console.data.db.AgentSessionRow

/**
 * Sidebar session ordering — pure port of src/components/agent/session-tree.ts.
 * Sessions cluster by cwd into a group tree (ordered within + across groups by
 * the persisted flat `sessionOrder`), with fork lineage nesting each fork right
 * after its parent. Al is excluded here (the list pins it first separately).
 */

data class SessionGroupNode(
    val cwd: String,          // "" = "(no directory)" bucket
    val label: String,        // path segment relative to parent
    val fullPath: String,
    val sessions: List<AgentSessionRow>,
    val children: List<SessionGroupNode>,
    val depth: Int,
)

/** A flat, renderable row: a group header, or a session at some indent depth. */
sealed interface SidebarRow {
    data class Group(val node: SessionGroupNode) : SidebarRow
    data class Session(val session: AgentSessionRow, val depth: Int) : SidebarRow
}

private fun dirBasename(path: String): String {
    val parts = path.trimEnd('/').split("/")
    return parts.lastOrNull()?.ifEmpty { path } ?: path
}

fun buildGroupTree(sessions: List<AgentSessionRow>, order: List<String>): List<SessionGroupNode> {
    val orderIdx = order.withIndex().associate { (i, id) -> id to i }
    fun sessionKey(s: AgentSessionRow) = orderIdx[s.id] ?: Int.MAX_VALUE
    val cmp = Comparator<AgentSessionRow> { a, b ->
        val ai = sessionKey(a); val bi = sessionKey(b)
        if (ai != bi) ai - bi else (b.createdAt - a.createdAt).let { if (it > 0) 1 else if (it < 0) -1 else 0 }
    }

    val byCwd = LinkedHashMap<String, MutableList<AgentSessionRow>>()
    for (s in sessions) byCwd.getOrPut(s.cwd ?: "") { mutableListOf() }.add(s)
    for (list in byCwd.values) list.sortWith(cmp)

    val cwds = byCwd.keys.sortedBy { it.length }
    val nodeByCwd = LinkedHashMap<String, MutableNode>()
    for (cwd in cwds) nodeByCwd[cwd] = MutableNode(cwd, if (cwd.isEmpty()) "(no directory)" else dirBasename(cwd), cwd, byCwd[cwd]!!, mutableListOf(), 0)

    val roots = mutableListOf<MutableNode>()
    for (cwd in cwds) {
        val node = nodeByCwd[cwd]!!
        var parentCwd: String? = null
        if (cwd.isNotEmpty()) {
            for (other in cwds) {
                if (other.isEmpty() || other == cwd) continue
                if (cwd.startsWith("$other/") && (parentCwd == null || other.length > parentCwd!!.length)) parentCwd = other
            }
        }
        if (parentCwd != null) {
            val parent = nodeByCwd[parentCwd]!!
            parent.children.add(node)
            node.depth = parent.depth + 1
            node.label = cwd.substring(parentCwd!!.length + 1)
        } else roots.add(node)
    }

    fun groupKey(n: MutableNode): Int =
        if (n.sessions.isNotEmpty()) sessionKey(n.sessions.first())
        else if (n.children.isNotEmpty()) groupKey(n.children.first()) else Int.MAX_VALUE
    fun sortRec(arr: MutableList<MutableNode>) { arr.sortBy { groupKey(it) }; for (n in arr) sortRec(n.children) }
    sortRec(roots)
    return roots.map { it.freeze() }
}

private class MutableNode(
    val cwd: String, var label: String, val fullPath: String,
    val sessions: MutableList<AgentSessionRow>, val children: MutableList<MutableNode>, var depth: Int,
) {
    fun freeze(): SessionGroupNode = SessionGroupNode(cwd, label, fullPath, sessions, children.map { it.freeze() }, depth)
}

/** If a single root shares one cwd, drop its redundant header — promote its
 *  sessions to the top level and its children become roots. */
fun peelUniversalRoot(roots: List<SessionGroupNode>): Pair<List<AgentSessionRow>, List<SessionGroupNode>> {
    if (roots.size != 1) return emptyList<AgentSessionRow>() to roots
    val only = roots[0]
    fun shift(n: SessionGroupNode, d: Int): SessionGroupNode = n.copy(depth = n.depth + d, children = n.children.map { shift(it, d) })
    return only.sessions to only.children.map { shift(it, -1) }
}

/** Arrange sessions (all one group) into fork lineage: each fork right after
 *  its parent, one indent deeper. */
fun arrangeLineage(sessions: List<AgentSessionRow>): List<Pair<AgentSessionRow, Int>> {
    val inSet = sessions.mapNotNull { it.claudeSessionId }.toSet()
    val childrenOf = HashMap<String, MutableList<AgentSessionRow>>()
    for (s in sessions) {
        val p = s.parentClaudeSessionId
        if (p != null && p in inSet) childrenOf.getOrPut(p) { mutableListOf() }.add(s)
    }
    val out = mutableListOf<Pair<AgentSessionRow, Int>>()
    fun emit(s: AgentSessionRow, depth: Int) {
        out.add(s to depth)
        s.claudeSessionId?.let { csid -> childrenOf[csid]?.forEach { emit(it, depth + 1) } }
    }
    for (s in sessions) {
        val isRoot = s.parentClaudeSessionId == null || s.parentClaudeSessionId !in inSet
        if (isRoot) emit(s, 0)
    }
    return out
}

/** Flatten the group tree into renderable rows (headers + indented sessions),
 *  skipping collapsed groups' contents. Al excluded (caller prepends it). */
fun flattenSidebar(
    sessions: List<AgentSessionRow>,
    order: List<String>,
    collapsed: Set<String>,
): List<SidebarRow> {
    val (rootSessions, roots) = peelUniversalRoot(buildGroupTree(sessions, order))
    val out = mutableListOf<SidebarRow>()
    for ((s, d) in arrangeLineage(rootSessions)) out.add(SidebarRow.Session(s, d))
    fun walk(nodes: List<SessionGroupNode>) {
        for (n in nodes) {
            out.add(SidebarRow.Group(n))
            if (n.cwd in collapsed) continue
            for ((s, d) in arrangeLineage(n.sessions)) out.add(SidebarRow.Session(s, d))
            walk(n.children)
        }
    }
    walk(roots)
    return out
}
